Unconditionally add asserts after export

tugsbayasgalan opened the pull request:
Summary: Today AOTAutograd drops some of assert nodes so we reapply it after strict export.

Test Plan: CI

Reviewed By: angelayi

Differential Revision: D57786907

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -194,6 +206,14 @@ def go(node, keypath):
                             and isinstance(keypath[0], CallMethodKey)
                             and isinstance(keypath[1], pytree.SequenceKey)
                         ):
+                            if keypath[0].name == "size":
```
For completeness, you should do `stride` and (not here, but below) `storage_offset` as well

ezyang commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -48,6 +48,18 @@ def insert_deferred_runtime_asserts(
+            and node.target == torch.ops.aten.sym_constrain_range.default
+        ):
+            assert len(node.args) == 1
+            nodes_that_already_have_sym_constraint_range.add(node.args[0])
```
I'm a bit ambivalent about this. In particular, if the ShapeEnv has a more refined value range, then this test could potentially prevent us from putting the more accurate value range in the graph (since you are not inspecting the contents of this call.)

Now, what should we do about the duplication problem? One approach for this problem is to lean into one of the odd things I noticed while I was working on Inductor: it's actually quite hard to stop Inductor from DCE'ing calls to `sym_constrain_range` and `assert_scalar`. The typical course of events goes like this:

1. We have a rule that says if someone does a `torch._check(True)`, we can drop it
2. We perform simplifications on sympy expressions based on things in the guard environment.
3. When I run `torch._check(x == 4)`, this adds `x == 4` to the list of guards
4. Then, when I see if I can simplify `x == 4`, this simplifies to `True`, because it's already in the guard environment
5. Oops, bye bye torch._check(True)!

So basically, you want to try to NOT put any checks into the graph initially, and rely on the runtime asserts pass to actually go ahead and insert them.

IDK, while I don't like the code here, I'm open to many ways to solve it. Maybe this gives you some ideas.

ezyang replied:
The easiest way forward is to not do this dedupe, and suck it up for the duplicate asserts lol

tugsbayasgalan replied:
Hmm it looks like when we add sym_constrain_range in the graph, we always simplify the min and max value to be concrete integers. Why not we just cache the tuple of (node_name, min, max)?

ezyang replied:
I guess you can do a one off optimization like that too

ezyang commented:
Looks good except for the unsound removal of duplicate asserts

tugsbayasgalan committed:
Unconditionally add asserts after export (#127132)

Summary:

Today AOTAutograd drops some of assert nodes so we reapply it after strict export.

Test Plan: CI

Reviewed By: angelayi, zhxchen17

Differential Revision: D57786907
